### PR TITLE
[Feature] benchmark can add work_dir and repeat times, test.py now has default work-dir

### PR DESCRIPTION
**Pull request by RockeyCoss**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

To deal with the fluctuation of fps,  we need to run the benchmark.py several times and use the average of several evaluations as the result. Therefore, I modified the benchmark.py. Now you can specify how many times you want to evaluate using parameter --repeat-times. What's more. You can specify the work dir by using --work-dir to dump the evaluation results into the directory you specified as json.

What's more, referring to train.py, now test.py will set work-dir automatically when work-dir is not supplied.

## Modification

Add parameter --work-dir and --repeat-times

### CODE REVIEW

File: tools/benchmark.py [hunk 1]
```diff
@@ -25,61 +33,81 @@ def main():
     args = parse_args()
 
     cfg = Config.fromfile(args.config)
+    if args.work_dir is not None:
+        mmcv.mkdir_or_exist(osp.abspath(args.work_dir))
+        timestamp = time.strftime('%Y%m%d_%H%M%S', time.localtime())
+        json_file = osp.join(args.work_dir, f'fps_{timestamp}.json')
```
Comment (Junjun2016) [inline tools/benchmark.py:39]:
```suggestion
        json_file = osp.join(args.work_dir, args.config， f'fps_{timestamp}.json')
```

It is better to add the config name.

File: tools/benchmark.py [hunk 2]
```diff
@@ -25,61 +33,81 @@ def main():
...
+            dist=False,
+            shuffle=False)
+
+        # build the model and load checkpoint
+        cfg.model.train_cfg = None
+        model = build_segmentor(cfg.model, test_cfg=cfg.get('test_cfg'))
+        fp16_cfg = cfg.get('fp16', None)
+        if fp16_cfg is not None:
+            wrap_fp16_model(model)
+        load_checkpoint(model, args.checkpoint, map_location='cpu')
```
Comment (Junjun2016) [inline tools/benchmark.py:66]: If the `args.checkpoint` does not exist or is given, we could skip this step.
Refer to ServerPerf.

**Review by MengzhangLI — APPROVED** [review]

File: tools/benchmark.py [hunk 3]
```diff
@@ -13,10 +16,15 @@
 
 def parse_args():
     parser = argparse.ArgumentParser(description='MMSeg benchmark a model')
-    parser.add_argument('config', help='test config file path')
+    parser.add_argument('config', help='test config file path', required=True)
     parser.add_argument('checkpoint', help='checkpoint file')
```
Comment (Junjun2016) [inline tools/benchmark.py:20]: Make it to be an optional argument.

**Review by Junjun2016 — APPROVED** [review]

File: tools/benchmark.py [hunk 4]
```diff
@@ -25,61 +33,88 @@ def main():
...
+                print(f'Overall fps: {fps:.2f} img / s')
+                print()
+                benchmark_dict[f'overall_fps_{time_index + 1}'] = round(fps, 2)
+                overall_fps_list.append(fps)
+                break
+    benchmark_dict['average_fps'] = round(np.mean(overall_fps_list), 2)
+    benchmark_dict['fps_variance'] = round(np.var(overall_fps_list), 4)
+    print(f'Average fps of {repeat_times} evaluations: '
+          f'{benchmark_dict["average_fps"]}')
+    print(f'The variance of {repeat_times} evalutations:'
```
Comment (MeowZheng) [inline tools/benchmark.py:115]: typo evalutations -> evaluations

File: tools/benchmark.py [hunk 5]
```diff
@@ -17,6 +20,11 @@ def parse_args():
     parser.add_argument('checkpoint', help='checkpoint file')
     parser.add_argument(
         '--log-interval', type=int, default=50, help='interval of logging')
+    parser.add_argument(
+        '--work-dir',
+        help=('if specified, the results will be dumped'
+              'into the directory as json'))
```
Comment (MeowZheng) [inline tools/benchmark.py:26]:
```suggestion
        help=('if specified, the results will be dumped '
              'into the directory as json'))
```

File: tools/benchmark.py [hunk 6]
```diff
@@ -25,61 +33,88 @@ def main():
...
                 fps = (i + 1 - num_warmup) / pure_inf_time
-                print(f'Done image [{i + 1:<3}/ {total_iters}], '
-                      f'fps: {fps:.2f} img / s')
-
-        if (i + 1) == total_iters:
-            fps = (i + 1 - num_warmup) / pure_inf_time
-            print(f'Overall fps: {fps:.2f} img / s')
-            break
+                print(f'Overall fps: {fps:.2f} img / s')
+                print()
```
Comment (MeowZheng) [inline tools/benchmark.py:107]:
```suggestion
                print(f'Overall fps: {fps:.2f} img / s\n')
```

**Review by MeowZheng — APPROVED** [review]

**Pull request merged**